src/blocker.rs

```diff
@@ -265,18 +247,95 @@ impl Blocker {
             }
         });
 
+        let important = filter.is_some() && filter.as_ref().map(|f| f.is_important()).unwrap_or_else(|| false);
+
+        let rewritten_url = if important {
+            None
+        } else {
+            Self::apply_removeparam(&self.removeparam, request, request_tokens)
+        };
+
         // If something has already matched before but we don't know what, still return a match
         let matched = exception.is_none() && (filter.is_some() || matched_rule);
         BlockerResult {
             matched,
-            important: filter.is_some() && filter.as_ref().map(|f| f.is_important()).unwrap_or_else(|| false),
+            important,
             redirect,
+            rewritten_url,
             exception: exception.as_ref().map(|f| f.to_string()), // copy the exception
             filter: filter.as_ref().map(|f| f.to_string()),       // copy the filter
             error: None,
         }
     }
 
+    fn apply_removeparam(removeparam_filters: &NetworkFilterList, request: &Request, request_tokens: lifeguard::Recycled<Vec<u64>>) -> Option<String> {
+        /// Represents an `&`-separated argument from a URL query parameter string
+        enum QParam<'a> {
+            /// Just a key, e.g. `...&key&...`
+            KeyOnly(&'a str),
+            /// Key-value pair separated by an equal sign, e.g. `...&key=value&...`
+            KeyValue(&'a str, &'a str),
+        }
+
+        impl<'a> std::fmt::Display for QParam<'a> {
+            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                match self {
+                    Self::KeyOnly(k) => write!(f, "{}", k),
+                    Self::KeyValue(k, v) => write!(f, "{}={}", k, v),
+                }
+            }
+        }
+
+        let url = &request.original_url;
+        // Only check for removeparam if there's a query string in the request URL
+        if let Some(i) = url.find('?') {
+            // String indexing safety: indices come from `.len()` or `.find(..)` on individual
+            // ASCII characters (1 byte each), some plus 1.
+            let params_start = i + 1;
+            let hash_index = if let Some(j) = url[params_start..].find('#') { params_start + j } else { url.len() };
+            let qparams = &url[params_start..hash_index];
+            let mut params: Vec<(QParam, bool)> = qparams
+                .split('&')
+                .map(|pair| {
+                    if let Some((k, v)) = pair.split_once('=') {
+                        QParam::KeyValue(k, v)
+                    } else {
+                        QParam::KeyOnly(pair)
+                    }
+                })
+                .map(|param| (param, true))
+                .collect();
+
+            let filters = removeparam_filters.check_all(request, &request_tokens, &NO_TAGS);
+            let mut rewrite = false;
+            for removeparam_filter in filters {
```

Comment:
As an optimization (maybe), we could avoid looping through the params that have been set to `include = false`.

Reply:
It's doing such a small amount of work per iteration; I'd be more concerned about the ensuing branch predictor misses than the saved cycles. Hard to say without benchmarking of course, but I don't think it's worth it for this small loop :sweat_smile:

Reply:
Yeah, I was just thinking that let's say we're checking `https://example.com/?fbclid=123`, then we'll remove the param as soon as we hit the fbclid rule (i.e. clearing the query string entirely), but we'd still be checking that soon-to-be-removed param against the other `filters`. Seems like we could exit early.

Now that I think more about it though, we could be hitting exception filters later on though that would bring this param back right?

```
||*$removeparam=fbclid
...
@@example.com$removeparam=mkt_tok
```